### PR TITLE
Using local storage for notification toggling

### PR DESCRIPTION
**Pull request by dvoytenko**

### CODE REVIEW

File: extensions/amp-user-notification/0.1/amp-user-notification.js [hunk 1]
```diff
@@ -133,7 +150,7 @@ export class AmpUserNotification extends AMP.BaseElement {
    * @private
    */
   buildGetHref_(ampUserId) {
-    return this.urlReplacements_.expand(this.showIfHref_).then(href => {
+    return this.urlReplacements_.expand(assert(this.showIfHref_)).then(href => {
```
Comment (erwinmombay) [inline extensions/amp-user-notification/0.1/amp-user-notification.js]: is it possible to do the assert earlier than this?

File: src/cookies.js [hunk 2]
```diff
@@ -27,7 +27,8 @@
  * @return {?string}
  */
 export function getCookie(win, name) {
-  const cookieString = win.document.cookie;
+  const cookieString = window.AMP_TEST && window.AMP_TEST_COOKIE ?
```
Comment (erwinmombay) [inline src/cookies.js]: can we add an explanation for amp_test_cookie

File: extensions/amp-user-notification/0.1/amp-user-notification.js (hunk 1 above)
Comment (dvoytenko) [inline extensions/amp-user-notification/0.1/amp-user-notification.js]: It is done early as well. Just being attentive to types.

File: src/cookies.js (hunk 2 above)
Comment (dvoytenko) [inline src/cookies.js]: Hmm. Kind of don't want it in the public API...

**Comment by dvoytenko** [comment]
@erwinmombay added tests as well. ptal.

File: extensions/amp-user-notification/0.1/amp-user-notification.js [hunk 3]
```diff
@@ -104,20 +106,35 @@ export class AmpUserNotification extends AMP.BaseElement {
...
-        'should have "data-dismiss-href" attribute.');
+    /** @private @const {?string} */
+    this.storageKey_ = isExperimentOn(this.getWin(), 'amp-storage') ?
+        'amp-user-notification:' + this.elementId_ : null;
 
-    /** @private @const {string} */
-    this.dismissHref_ = assertHttpsUrl(
-        this.element.getAttribute('data-dismiss-href'), this.element);
+    /** @private @const {?string} */
+    this.showIfHref_ = this.element.getAttribute('data-show-if-href');
```
Comment (erwinmombay) [inline extensions/amp-user-notification/0.1/amp-user-notification.js]: can we either add a note here or on the PR to update validator. (or create an issue)

Comment (dvoytenko) [inline extensions/amp-user-notification/0.1/amp-user-notification.js]: I will, as soon as the PR is approved by @erwinmombay and @cramforce

File: extensions/amp-user-notification/0.1/amp-user-notification.js [hunk 4]
```diff
@@ -215,6 +232,36 @@ export class AmpUserNotification extends AMP.BaseElement {
...
+          // Consent has been accepted. Nothing more to do.
+          return false;
+        }
+        if (this.showIfHref_) {
+          // Ask remote endpoint if available.
+          return this.shouldShowViaXhr_();
+        }
+        // Otherwise, show the notification.
+        return true;
+      }, reason => {
```
Comment (erwinmombay) [inline extensions/amp-user-notification/0.1/amp-user-notification.js]: can we switch this to `.catch` instead, easier to read.

**Comment by erwinmombay** [comment]
@dvoytenko looks great, LGTM

File: extensions/amp-user-notification/0.1/amp-user-notification.js (hunk 4 above)
Comment (dvoytenko) [inline extensions/amp-user-notification/0.1/amp-user-notification.js]: Done

**Comment by dvoytenko** [comment]
Assigning to @cramforce for an additional review.

File: extensions/amp-user-notification/0.1/amp-user-notification.js [hunk 5]
```diff
@@ -104,23 +106,40 @@ export class AmpUserNotification extends AMP.BaseElement {
...
+    if (!this.storageKey_) {
+      assert(this.showIfHref_,
+          `"amp-user-notification" (${this.elementId_}) ` +
+          'should have "data-show-if-href" attribute.');
+      assert(this.dismissHref_,
+          `"amp-user-notification" (${this.elementId_}) ` +
+          'should have "data-dismiss-href" attribute.');
+    }
+
+    if (!this.win_.AMP_TEST) {
```
Comment (cramforce) [inline extensions/amp-user-notification/0.1/amp-user-notification.js]: Why not in tests? Comment please

File: extensions/amp-user-notification/0.1/amp-user-notification.js [hunk 6]
```diff
@@ -104,23 +106,40 @@ export class AmpUserNotification extends AMP.BaseElement {
...
         'amp-user-notification should have an id.');
 
-    assert(this.element.hasAttribute('data-show-if-href'),
-        `"amp-user-notification" (${this.elementId_}) ` +
-        'should have "data-show-if-href" attribute.');
-    /** @private @const {string} */
-    this.showIfHref_ = assertHttpsUrl(
-        this.element.getAttribute('data-show-if-href'), this.element);
+    /** @const @private {!Promise<!Storage>} */
+    this.storagePromise_ = storageFor(this.win_);
```
Comment (cramforce) [inline extensions/amp-user-notification/0.1/amp-user-notification.js]: Is this guaranteed to succeed? What if `amp-analytics` is not loaded?

File: extensions/amp-user-notification/0.1/amp-user-notification.js [hunk 7]
```diff
@@ -104,23 +106,40 @@ export class AmpUserNotification extends AMP.BaseElement {
...
+    /** @private @const {?string} */
+    this.storageKey_ = isExperimentOn(this.getWin(), 'amp-storage') ?
+        'amp-user-notification:' + this.elementId_ : null;
 
-    /** @private @const {string} */
-    this.dismissHref_ = assertHttpsUrl(
-        this.element.getAttribute('data-dismiss-href'), this.element);
+    /** @private @const {?string} */
+    this.showIfHref_ = this.element.getAttribute('data-show-if-href');
+    if (this.showIfHref_) {
```
Comment (cramforce) [inline extensions/amp-user-notification/0.1/amp-user-notification.js]: Make an either/or assert: has href or storage?

File: src/cookies.js [hunk 8]
```diff
@@ -27,7 +27,8 @@
  * @return {?string}
  */
 export function getCookie(win, name) {
-  const cookieString = win.document.cookie;
+  const cookieString = win.AMP_TEST && win.AMP_TEST_COOKIE ?
```
Comment (cramforce) [inline src/cookies.js]: Not a fan.

File: extensions/amp-user-notification/0.1/amp-user-notification.js (hunk 5 above)
Comment (dvoytenko) [inline extensions/amp-user-notification/0.1/amp-user-notification.js]: Removed. Found another way to do this. But ultimately, I want unit-tests to be untainted by service integration - it's tested separately.

File: extensions/amp-user-notification/0.1/amp-user-notification.js (hunk 6 above)
Comment (dvoytenko) [inline extensions/amp-user-notification/0.1/amp-user-notification.js]: There's code that protects from failures here. It's tested as well.

File: extensions/amp-user-notification/0.1/amp-user-notification.js (hunk 7 above)
Comment (dvoytenko) [inline extensions/amp-user-notification/0.1/amp-user-notification.js]: Yes, it does this few lines down.

File: src/cookies.js (hunk 8 above)
Comment (dvoytenko) [inline src/cookies.js]: Agree. Removed.

**Comment by dvoytenko** [comment]
@cramforce  changes are complete. PTAL.

**Comment by cramforce** [comment]
LGTM

File: extensions/amp-user-notification/0.1/amp-user-notification.js [hunk 9]
```diff
@@ -235,7 +289,16 @@ export class AmpUserNotification extends AMP.BaseElement {
     this.element.classList.remove('amp-active');
     this.element.classList.add('amp-hidden');
     this.dialogResolve_();
-    this.postDismissEnpoint_();
+
+    // Store and post.
+    if (this.storageKey_) {
+      this.storagePromise_.then(storage => {
```
Comment (jridgewell) [inline extensions/amp-user-notification/0.1/amp-user-notification.js]: Shouldn't this be done as a `then` callback inside the `storageKey_` conditional in `#shouldShow`?

Comment (dvoytenko) [inline extensions/amp-user-notification/0.1/amp-user-notification.js]: Sorry?

Comment (jridgewell) [inline extensions/amp-user-notification/0.1/amp-user-notification.js]: My bad, looking at the compressed diff view I thought this was in `#shouldShowViaXhr_`. Ignore my comments.

**Pull request merged**